Enhance semantic conventions for HTTP

Oberon00 opened the pull request:
This PR aims to improve the semantic conventions for HTTP. For that, it also draws from experience at Dynatrace. We plan to later (hopefully soon) make similar PRs for other semantic convention areas, including a clean-up of the host.* and peer.* attributes used here.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
-Given an [RFC 3986](https://www.ietf.org/rfc/rfc3986.txt) compliant URI of the form
-`scheme:[//authority]path[?query][#fragment]`, the span name of the span SHOULD
-be set to to the URI path value.
+These span types represent HTTP requests. They can be used for http and https
```
```suggestion
`HTTP` span types (client and server) represent HTTP requests. They can be used for http and https
```

Oberon00 replied:
I reworded this.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
-| `http.url` | HTTP URL of this request, represented as `scheme://host:port/path?query#fragment` E.g. `"https://example.com:779/path/12314/?q=ddds#123"`. | Yes |
-| `http.status_code` | [HTTP response status code](https://tools.ietf.org/html/rfc7231). E.g. `200` (integer) | No |
-| `http.status_text` | [HTTP reason phrase](https://www.ietf.org/rfc/rfc2616.txt). E.g. `"OK"` | No |
+| `http.url` | Full HTTP request URL in the form `scheme://host[:port]/path?query[#fragment]`. Usually the fragment is not transmitted over HTTP, but if it is known, it should be included nevertheless. | Defined later. |
```
why to mention it in a common section if it is defined later?

SergeyKanzhelev replied:
Also for servers I understand you propose to make it a least preferable option. So perhaps define it for each type

Oberon00 replied:
The "Defined later" is in the "Required?" column. Thus only whether it is required is defined later, the actual definition of what should be put there is the same for client and server (though you are right that different nuances apply). I like keeping documentation DRY like code and didn't want to repeat that ` scheme://...` part all over.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `http.url` | Full HTTP request URL in the form `scheme://host[:port]/path?query[#fragment]`. Usually the fragment is not transmitted over HTTP, but if it is known, it should be included nevertheless. | Defined later. |
+| `http.status_code` | [HTTP response status code][]. E.g. `200` (integer) | No |
+| `http.status_text` | [HTTP reason phrase][]. E.g. `"OK"` | No |
+| `http.flavor` | Kind of HTTP protocol used: `"1.0"`, `"1.1"`, `"2"`, `"SPDY"` or `"QUIC"`. |  If not TCP-based (`QUIC`). |
```
why is it required? Look like an overkill for the most use cases.

Oberon00 replied:
I suggested to require it only if the protocol is not TCP-based because that influences the meaning of any port information (Is it TCP/80 or UDP/80?). But you are right this is probably too strict. I'll change it to not required. 👌

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `http.flavor` | Kind of HTTP protocol used: `"1.0"`, `"1.1"`, `"2"`, `"SPDY"` or `"QUIC"`. |  If not TCP-based (`QUIC`). |
 
-## HTTP server
+It is recommended to also use the `peer.*` attributes, especially `peer.ip*`.
```
so for client type - should one use `peer.host` or `http.host`? I'd suggest to expand on specific peer attributes that are required.

As mentioned here https://github.com/open-telemetry/opentelemetry-specification/issues/245 for DB attributes - `peer.service` may carry duplication value

Oberon00 replied:
`http.host` is not even defined for clients. It can be inferred trivially from the URL, which is typically what is available when you instrument the client side (the Host header is then typically set by the HTTP library somewhere in the implementation).

But if you wanted you could define the distinction like this: `http.host`  is what ends up in the HTTP request header whereas `peer.host`  is what is used to look up the IP address with which the underlying TCP connection is established (which might be different in proxying scenarios).

Oberon00 replied:
Do you think I should add something to that effect in the text?

tedsuo replied:
@Oberon00 I suggest that the `peer` object be addressed in a future PR. Its current form is vague and non-standard, which has raised some problems for a while. 

I suggest that peer be eliminated, and split up as follows:
* application-level fields move to application objects - `db.hostname`, `db.address`, etc, so there is no confusing partial overlap with http concepts.
* transport and internet layer objects - `tcp`, `udp`, `ipv4`, `ipv6`, etc

This could be done as part of a move to a more nested approach to semantic data. So, top level `http`, `tcp`, `ip`, and `db` objects are considered part of `self`. The same objects nested under `peer` describe the other side of the connection.

Again, I don't think that needs to be addressed part of this PR, but I wanted to jot down my reasoning here.

Oberon00 replied:
Moving this to #295

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
 
-Given an inbound request for a route (e.g. `"/users/:userID?"` the `name`
-attribute of the span SHOULD be set to this route.
+`http.url` is required and represents the HTTP URL used to (initially) make this request.
```
would we prefer the same scheme of preference of url parts when/if they are readily available?

Oberon00 replied:
You commented on the client side markdown here but your comment seems to apply to the server side. Please clarify.

Oberon00 replied:
Ah, I understand what you meant now!

I think we should set the span attributes to the most "original" data, and the Host and target are usually derived from the URL at the client side, while it is the other way round on the server side.

SergeyKanzhelev replied:
should we tell that if other data is readily available, one must prefer the separate parts scheme than? Let's say if the library sets the host once and use it for multiple requests providing only the path part.

Oberon00 replied:
That makes sense. I'll move http.host and http.target to the general attributes.

Oberon00 replied:
Done.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
 
-Given an inbound request for a route (e.g. `"/users/:userID?"` the `name`
-attribute of the span SHOULD be set to this route.
+`http.url` is required and represents the HTTP URL used to (initially) make this request.
```
Can you please elaborate on "initially"

SergeyKanzhelev replied:
perhaps add a note with the scenario

Oberon00 replied:
The initially is meant to cover the case where you make a request that gets redirected. E.g. you request example.com/foo but are 302'd to example.com/bar. What should you put in `http.url`? The "initially" makes it clear: The `https://example.com/foo`  you initially wanted to request.

Oberon00 replied:
I clarified that in the text.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+
+For an HTTP server span, `SpanKind` MUST be `Server`.
+
+Given an inbound request for a route (e.g. `"/users/:userID?"` the `name` attribute of the span SHOULD be set to this route. If the route does not include the application root path, it SHOULD be prepended to the span name.
```
please explain the application root path concept

Oberon00 replied:
Would a reference to "see the `http.app_root` attribute" be good too?

Oberon00 replied:
I added a "Definitions" section that explains the concept (and others).

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+
+Given an inbound request for a route (e.g. `"/users/:userID?"` the `name` attribute of the span SHOULD be set to this route. If the route does not include the application root path, it SHOULD be prepended to the span name.
+
+If the route cannot be determined, the `name` attribute MUST be set as defined in the general semantic conventions for HTTP.
```
general semantic convention requires to take a path from the Url. Later in the doc you re saying that url and/or path may not be always available. So this statement may raise some confusions

Oberon00 replied:
Yeah, I noticed that too but in that case, the route simply is "another value that represents the identity of the request and has a lower cardinality", so its just a clarification of the general case for the server scenario, no contradiction.

Oberon00 replied:
Probably after #270 this will need a rewording.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
-| `http.status_code` | [HTTP response status code](https://tools.ietf.org/html/rfc7231). E.g. `200` (integer) | No |
-| `http.status_text` | [HTTP reason phrase](https://www.ietf.org/rfc/rfc2616.txt). E.g. `"OK"` | No |
+| `http.target` | The full request target as passed in a [HTTP request line][] or equivalent, e.g. `/path/12314/?q=ddds#123"`. | [1] |
+| `http.host` | The value of the [HTTP host header][]. Note that this might be empty or not present. | [1] |
```
Should it be carried as empty value when no header present or attribute must be omitted?

Oberon00 replied:
Good point, I thought about that too but did not think it was worth specifying that detail. If we wanted to distinguish all cases:

* Not set if the span creator simply did not care to set it, does not know its value.
* Empty string if it is actually an empty string.
* The boolean `false` (I'd prefer a special null-value, but we don't have that) if the header is actually not set.

Oberon00 replied:
Do you think I should add that? Or should I leave it unspecified?

tedsuo replied:
It seems fine to not set it, rather than expend the overhead to set it to false.

Oberon00 replied:
I added the sentence

>When the header is empty or not present, this attribute should be the same.

With named tracers, one should also be able to take an estimated guess on whether the attribute is missing because the instrumentation did not care about it or because the header was really not present.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `http.target` | The full request target as passed in a [HTTP request line][] or equivalent, e.g. `/path/12314/?q=ddds#123"`. | [1] |
+| `http.host` | The value of the [HTTP host header][]. Note that this might be empty or not present. | [1] |
+| `http.scheme` | The URI scheme identifying the used protocol: `"http"` or `"https"` | [1] |
+| `http.server_name` | The (primary) server name (usually not including a port). This should be obtained via configuration, e.g. the Apache [`ServerName`][ap-sn] or NGINX [`server_name`][nx-sn] directive. If no such configuration can be obtained, this attribute MUST NOT be set ( `host.name` should be used instead). | [1] |
```
Logically this attribute belongs to the `Resource`, not in every single http request information. Is it only needed when `host` wasn't specified?

Oberon00 replied:
When you talk about `host`, do you mean http.host or host.name? For when it is needed, see [1] below, and also the note on `http.host` and why a `server_name`  is useful even when `http.host`  is given.

Oberon00 replied:
See also #274

Oberon00 replied:
I think this is clear now in the current revision of the text if you read the "Definitions" section.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `http.target` | The full request target as passed in a [HTTP request line][] or equivalent, e.g. `/path/12314/?q=ddds#123"`. | [1] |
+| `http.host` | The value of the [HTTP host header][]. Note that this might be empty or not present. | [1] |
+| `http.scheme` | The URI scheme identifying the used protocol: `"http"` or `"https"` | [1] |
+| `http.server_name` | The (primary) server name (usually not including a port). This should be obtained via configuration, e.g. the Apache [`ServerName`][ap-sn] or NGINX [`server_name`][nx-sn] directive. If no such configuration can be obtained, this attribute MUST NOT be set ( `host.name` should be used instead). | [1] |
```
based on this definition `host` is less prefered than `server_name`. Should it be other way around and when `host` present - `server_name` MUST not be collected?

Oberon00 replied:
Again, please clarify whether you mean `host.name`  or `http.host`. Assuming you mean `http.host`, see the note on its unreliability below. If you meant `host.name`, that can be something of questionable usability like `k8s-a34b4-k434334d-myservice.local`. Also a single host computer can host multiple server_name's, even in the same process.

Oberon00 replied:
Is this clearer with the new "Definitions" section?

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `http.host` | The value of the [HTTP host header][]. Note that this might be empty or not present. | [1] |
+| `http.scheme` | The URI scheme identifying the used protocol: `"http"` or `"https"` | [1] |
+| `http.server_name` | The (primary) server name (usually not including a port). This should be obtained via configuration, e.g. the Apache [`ServerName`][ap-sn] or NGINX [`server_name`][nx-sn] directive. If no such configuration can be obtained, this attribute MUST NOT be set ( `host.name` should be used instead). | [1] |
+| `host.name` |  Analogous to `peer.hostname` but for the host instead of the peer. | [1] |
```
Reading it I don't understand what should be the value here. Is it the same as `server_name`?

Oberon00 replied:
See my answer to your other comment: https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r329310425. `http.server_name` is a name specifically configured  and usually matched for that particular HTTP request. `host.name`  is in the worst case something you got from [`gethostname(2)`](http://man7.org/linux/man-pages/man2/gethostname.2.html).

Oberon00 replied:
I guess I should add a link to https://tools.ietf.org/html/rfc3875#section-4.1.14 or https://www.python.org/dev/peps/pep-3333/#environ-variables.

Oberon00 replied:
Explained more thoroughly now in the new Definitions subsection.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `host.name` |  Analogous to `peer.hostname` but for the host instead of the peer. | [1] |
+| `host.port` | Local port. E.g., `80` (integer). Analogous to `peer.port`. | [1] |
+| `http.route` | The matched route (path template). E.g. `"/users/:userID?"`. | No |
+| `http.app` | An identifier for the whole HTTP application. E.g. Flask app name, `spring.application.name`, etc. | No |
```
I'd suggest to keep it in resource API, not in individual requests

Oberon00 replied:
As I understand resources, they are process-wide. However, a process can host multiple apps.

Oberon00 replied:
I created #274 "Allow resources as span attributes"

bogdandrutu replied:
I am confused by this multiple "applications". What do you call an "application"?

Oberon00 replied:
Maybe the "Definitions" section above is helpful. But you may also consider the examples. The details depend on the technology. To add another example: In Java EE, an app is the entity described by the "web.xml" file and is defined as follows:

> A Web application is a collection of servlets, HTML pages, classes, and other resources that make up a complete application on a Web server.

(quote from SRV.9 of https://jcp.org/aboutJava/communityprocess/maintenance/jsr053/index2.html)

Oberon00 replied:
In Python, a application is defined by the [WSGI PEP](https://www.python.org/dev/peps/pep-3333/) as a callable with a certain behavior that is called by the webserver. For example, mod_wsgi implements the WSGI spec for apache and has a `WSGIScriptAlias` directive that maps a path prefix (context root) to a Python script with a WSGI application: https://modwsgi.readthedocs.io/en/develop/user-guides/quick-configuration-guide.html

Oberon00 replied:
If we want an abstract definition, maybe the application is "a collection of (compiled) code and static resources that is deployable to a web server". It is not a route. A typical application will consist of a multitude of routes.

Oberon00 replied:
@bogdandrutu I also added the Python and Java EE examples to the Definitions section. Please unresolve if you think this is still not clear.

SergeyKanzhelev replied:
So your suggestion on how to fix it is to append resource to a span. So if there are client spans of this span - will they inherit the resource? Wouldn't it be better for apps to register a resource for the entire Tracer used by anything inside this app?

I'd suggest to remove these properties from this document as it feels like they clearly belongs to the Resource and no matter how we will resolve association of a span to resource - it'e better to define them as labels on Resource.

Oberon00 replied:
If we decided that only one http app per process is supported, then the app would probably be already covered by the proposed `service.name` resource (https://github.com/open-telemetry/opentelemetry-specification/pull/303/files). But note that at the time I wrote this PR, there were no semantic conventions on resources at all, and I still think that resources are very vaguely specified, so it's hard for me to say whether something could be a resource or not, when I don't really know what a resource actually is.

SergeyKanzhelev replied:
I believe that Resource is not process-wide concept. It's per-Tracer concept. [Here](https://github.com/open-telemetry/opentelemetry-specification/blob/03018b1553c0cfee4a8d2bc497f018a4007b69d9/specification/sdk-resource.md#L1) it says:

> When used with distributed tracing, a resource can be associated with the TracerSdk. When associated with TracerSdk, all Spans produced by the Tracer, that is implemented by this TracerSdk, will automatically be associated with this Resource.

So for the case of multiple apps inside the process - each app may initialize it's own `Tracer` with the app-specific properties. Same properties may be interesting for the child spans so ideally this Tracer needs to be shared with all libraries used in this app.

jmacd replied:
(Don't forget that the `Meter` API will need equal treatment as far as resources go.)

Oberon00 replied:
Thanks, I will look into that! This also seems like something that needs to be reworded with named tracers (probably the Resources would be per TracerFactory).

Oberon00 replied:
I'll remove `http.app` for now and created #335 to track that.

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `host.port` | Local port. E.g., `80` (integer). Analogous to `peer.port`. | [1] |
+| `http.route` | The matched route (path template). E.g. `"/users/:userID?"`. | No |
+| `http.app` | An identifier for the whole HTTP application. E.g. Flask app name, `spring.application.name`, etc. | No |
+| `http.app_root` |The path prefix of the URL that identifies this `http.app`. Also known as "context root". If multiple roots exist, the one that was matched for this request should be used. | No |
```
Same, can it be set in resource API?

Oberon00 replied:
In addition to what I said in https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r329310518, this would require the back-end to redo the matching on the URL, if there are multiple roots (which may be acceptable though)

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+| `http.route` | The matched route (path template). E.g. `"/users/:userID?"`. | No |
+| `http.app` | An identifier for the whole HTTP application. E.g. Flask app name, `spring.application.name`, etc. | No |
+| `http.app_root` |The path prefix of the URL that identifies this `http.app`. Also known as "context root". If multiple roots exist, the one that was matched for this request should be used. | No |
+| `http.client_ip` | The IP address of the original client behind all proxies, if known (e.g. from [X-Forwarded-For][]). | No |
```
`peer.ip`?

Oberon00 replied:
Nope, `peer.ip4` or `peer.ip6`  would be the TCP-layer communication partner, which can be different from `http.client_ip` when a proxy (e.g. envoy) is in-between.

Oberon00 replied:
I added a sentence explaining the difference.

Oberon00 commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
+Given an [RFC 3986](https://tools.ietf.org/html/rfc3986) compliant URI of the form
+`scheme:[//host[:port]]path[?query][#fragment]`, the span name of the span SHOULD
+be set to to the URI path value, unless another value that represents the identity
+of the request and has a lower cardinality can be identified.
```
I just tried to make this sentence more clear, I had no intention of changing the meaning. You bring up a valid point though. Do you think I should address that in this PR?

Oberon00 replied:
Resolving as the duplicate at https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r329313767 has more discussion. Yuri also opened an issue #270

yurishkuro commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
-and has a lower cardinality than the URI path, this value MUST be used for the span name instead.
+Given an [RFC 3986](https://tools.ietf.org/html/rfc3986) compliant URI of the form
+`scheme:[//host[:port]]path[?query][#fragment]`, the span name of the span SHOULD
+be set to the URI path value, unless another value that represents the identity
```
-1

in OpenTracing we always explicitly recommended against doing that. In the REST services it is very common to have entity IDs as part of the path, e.g. `/api/user/{uuid}`. And since the instrumentation in general has no knowledge of whether the application's URL pattern includes entity IDs, recommending path as span name can easily lead to cardinality explosion.

Oberon00 replied:
First, I have not changed anything semantically about the span name here, so I'm not sure if this should be discussed here.

Second, What do you suggest instead for a generic HTTP instrumentation library?

Third, what are the goals and guarantees of a Span name anyway?

yurishkuro replied:
https://github.com/open-telemetry/opentelemetry-specification/issues/270

yurishkuro commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,103 @@ This way, the operator will not need to learn specifics of a language and
 
 | Attribute name | Notes and examples                                           | Required? |
 | :------------- | :----------------------------------------------------------- | --------- |
 | `component`    | Denotes the type of the span and needs to be `"http"`. | Yes |
```
what is the value of this attribute? The fact that this is an http span can be easily inferred from other attributes.

Oberon00 replied:
The component attribute exists in all conventions in this document. It is neither specific to the topic of this PR nor changed here. Please make a new PR or issue if you want to discuss that. I think removing the attribute just for HTTP wouldn't make sense.

yurishkuro replied:
https://github.com/open-telemetry/opentelemetry-specification/issues/271

reyang commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,132 @@ This way, the operator will not need to learn specifics of a language and
+
+|   Attribute name   |                                       Value                                       |
+| :----------------- | :-------------------------------------------------------------------------------- |
+| `component`        | `"http"`                                                                          |
```
Do we want `http.flavor` in this example?

Oberon00 replied:
We want, I'll add it.

Oberon00 replied:
I added it.

reyang commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,132 @@ This way, the operator will not need to learn specifics of a language and
+| `http.scheme`      | `"https"`                                                                         |
+| `http.route`       | `"/articles/:article_id"` (note that the `app_root` part is missing in this case) |
+| `http.status_code` | `200`                                                                             |
+| `http.status_text` | `"OK"`                                                                            |
```
Not sure if we plan to cover it in this PR - how do we interact with the [Span Status](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#set-status). Here are some examples:
* If an HTTP client span failed from client side (e.g. timeout, TLS handshake failure, connection reset, etc.), we might not have `http.status_code` and `http.status_text`, would `Span Status` be good enough to indicate the problem?
* If an HTTP client span _failed_ from the server side (HTTP 4xx and 5xx), what's the general guidance on [`Status`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#status)?

reyang replied:
Another example is the HTTP Server Redirection, for example, when the client is trying a access URI A, it got redirected to B, and then to C and back to A. The client might decide to give up due to circular redirection. From HTTP span's standpoint, all three HTTP client spans succeeded (with HTTP status = 301). However the last span might be considered as failed.

Oberon00 replied:
Good point, the HTTP conventions aren't up to date with the introduction of the Status field it seems. We should probably just reuse the relevant OpenCensus spec here: https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md#status

Oberon00 replied:
The redirect loop isn't covered there, but probably it should be DEADLINE_EXCEEDED.

Client errors can be distinguished from genuine HTTP errors by the absence of the (required) http status attribute. I'll also update the spec to clarify that status_code MUST NOT be set if no status was received from the server (i.e. in client/network error cases).

Oberon00 replied:
Having to always define these status code mappings is sort of a pain 😄

reyang replied:
We can leverage existing ones:
* https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#status
* https://github.com/grpc/grpc/blob/master/doc/statuscodes.md

Oberon00 replied:
BTW, this is #100

Oberon00 replied:
I added semantic conventions for the status, mostly the same ones as OpenCensus.

reyang commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,132 @@ This way, the operator will not need to learn specifics of a language and
+
+This span type represents an outbound HTTP request.
+
+For an HTTP client span, `SpanKind` MUST be `Client`.
```
We might need to cover #110.

Oberon00 replied:
Yeah, eventually we should add semantic conventions about how HTTP redirects should be handled. But I'd rather not do this in this PR.

tedsuo commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,195 @@ This way, the operator will not need to learn specifics of a language and
+### Name
+
+Given an [RFC 3986](https://tools.ietf.org/html/rfc3986) compliant URI of the form `scheme:[//host[:port]]path[?query][#fragment]`,
+the span name of the span SHOULD be set to the URI path value,
```
FWIW, in OpenTracing, we default to using the HTTP method as the default name, not the URI.
* Method is low cardinality. The default should be "too general," not "too specific," to avoid overwhelming systems by inserting high cardinality into a default index such as span name. Perhaps this matters less in the future.
* URI is often not available, making this an infeasible default. This is the real issue, besides cardinality, which forced us to pick a different default.

Note that backends can leverage the URI and other http properties for indexing and display purposes when it is available. Given how common HTTP operations are, we can presume that backends treat HTTP spans as a special case, so the generic name is not necessarily an issue.

tedsuo replied:
Ah, I see there already is an issue for this: https://github.com/open-telemetry/opentelemetry-specification/issues/270

Oberon00 replied:
As this PR is already getting pretty big, I'd rather not incorporate a resolution to that here.

Oberon00 commented:
@tigrannajaryan @SergeyKanzhelev @reyang I think I have addressed your comments and this PR has been lying around quite some time now already. If you have time, I'd like you to please re-review and possibly even give a ✔️ if you think it's fine. 😃

Oberon00 commented:
Thank you all for your feedback and reviews!

@ admins (@bogdandrutu, @SergeyKanzhelev):  I think this can be merged now.

jmacd commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,207 @@ This way, the operator will not need to learn specifics of a language and
+
+#### Definitions
+
+In web server configuration and web app deployment, multiple more or less common concepts have evolved that.
```
nit: Incomplete sentence "evolved that ...?"

Oberon00 replied:
I reworded that "introduction paragraph".

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,208 @@ This way, the operator will not need to learn specifics of a language and
+| `http.route` | The matched route (path template). E.g. `"/users/:userID?"`. | No |
+| `http.app` | An identifier for the whole HTTP application. E.g. Flask app name, `spring.application.name`, etc. | No |
+| `http.app_root` |The path prefix of the URL that identifies this `http.app`. If multiple roots exist, the one that was matched for this request should be used. | No |
+| `http.client_ip` | The IP address of the original client behind all proxies, if known (e.g. from [X-Forwarded-For][]). Note that this is not necessarily the same as `peer.ip*`, which would identify the nework-level peer, which may be a proxy. | No |
```
why `peer.ip` would not take an `X-Forwarded-For` into consideration? What would be the reason to have peer.ip carry the information that is meaningless semantically in many cases?

Oberon00 replied:
The good thing about having that `peer.ip` definition is that it is very simple (defined basically by a call to http://man7.org/linux/man-pages/man2/getpeername.2.html on the network socket) and applicable to all protocols. And it's not meaningless IMHO, as it at least allows you to detect that the request got through a proxy (for example, maybe you notice one day that all requests with a particular peer.ip take longer than other requests because the proxy has network troubles).

On the other hand http.client_ip is very http specific. The alternative would be to have peer.ip be defined differently for each type of span, which feels very unclean to me (And what if a span combines e.g. HTTP and some NoSQL DB and both have conflicting definitions of peer.ip? Now you have to pick one.)

SergeyKanzhelev commented on specification/data-semantic-conventions.md:
```diff
@@ -14,49 +14,208 @@ This way, the operator will not need to learn specifics of a language and
+| `http.method`      | `"GET"`                                                                           |
+| `http.flavor`      | `"1.1"`                                                                           |
+| `http.url`         | `"https://example.com:8080/webshop/articles/4?s=1"` (or not set)                  |
+| `http.target`      | `"/webshop/articles/4?s=1"`                                                       |
```
so this example is not how span will look like as you either have url or target, correct? Maybe it's worth splitting into two examples to avoid confusion

Oberon00 replied:
I only included a remark about the `http.url` value below the attribute table and removed the attribute from the table. I also added a client-side example.

Oberon00 commented:
@SergeyKanzhelev About your main objection regarding resources vs attributes, see https://github.com/open-telemetry/opentelemetry-specification/pull/263#discussion_r336907328

SergeyKanzhelev commented:
Would you be OK to move this forward by removing presumably resource labels into a separate PR? I feel strongly against those, but want to merge the rest to keep making progress

Oberon00 commented:
Which labels are you against in particular? Is it only the `http.app` attribute? If you really like, I think we can table that for now, even though it would be very interesting for Dynatrace.

But I would strongly prefer to keep at least all the fields that are (conditionally) required to (re)construct the URL (like http.server_name, host.name) in the Span because even if they would be moved to resources, it would mean that the back end has to access resources each time it wants to find the URL for an HTTP span which might not be as efficiently implementable. I think a bit of ["denormalization"](https://en.wikipedia.org/wiki/Denormalization) to speak in database terms is fine here. 

My main problem with moving things to resources is that I think they are really under-specified ATM and e.g. not implemented at all in Python. Also I'm not sure if existing protocols like Jaeger or Zipkin have a resource-like concept? AFAIK OpenTracing did not have it.

tigrannajaryan commented:
> Which labels/attributes are you against in particular? Is it only the `http.app` attribute? If you really like, I think we can table that for now, even though it would be very interesting for Dynatrace.

@Oberon00 `http.app` is unclear IMO. I'd advise to remove that from this PR to avoid delaying the whole PR because of it and work on it in an additional PR.

> 
> But I would strongly prefer to keep at least all the fields that are (conditionally) required to (re)construct the URL (like http.server_name, host.name) in the Span because even if they would be moved to resources, it would mean that the back end has to access resources each time it wants to find the URL for an HTTP span which might not be as efficiently implementable. I think a bit of ["denormalization"](https://en.wikipedia.org/wiki/Denormalization) to speak in database terms is fine here.

I believe we need be careful with this. You are right that it would be a denormalization of data and in my opinion that should be avoided. The semantic conventions are first of all a communication protocol to me. I believe a communication protocol should strive to for compactness, minimize data duplication and aim for normalized data also because it is typically costlier to re-normalize denormalized data than to de-normalize properly normalized data. 


> My main problem with moving things to resources is that I think they are really under-specified ATM and e.g. not implemented at all in Python. Also I'm not sure if existing protocols like Jaeger or Zipkin have a resource-like concept? AFAIK OpenTracing did not have it.

Jaeger has what I believe is an equivalent (or at least similar) to Resource - a Process:
https://github.com/jaegertracing/jaeger/blob/75888876df58f921aa976bd513ceb2704f47aacc/model/proto/model.proto#L92

SergeyKanzhelev commented:
I only was asking for `app` attribute.

Oberon00 commented:
@SergeyKanzhelev I removed `http.app` and created #335.

jmacd commented:
@Oberon00 In #303 I requested that we allow any key-value to be included as a resource or span attribute, for example. I think it's fine to spec `http.app` as a semantic convention, and if it's a constant across the process lifetime, it should be included in the resources. See: https://github.com/open-telemetry/opentelemetry-specification/pull/303#discussion_r337714858

tigrannajaryan commented:
> @Oberon00 In #303 I requested that we allow any key-value to be included as a resource or span attribute, for example. I think it's fine to spec `http.app` as a semantic convention, and if it's a constant across the process lifetime, it should be included in the resources. See: [#303 (comment)](https://github.com/open-telemetry/opentelemetry-specification/pull/303#discussion_r337714858)

@jmacd what happens if the Span specifies an attribute with a key that is already specified in the Resource associated with the Span? Do attributes in Span override attributes in the Resource semantically?

jmacd commented:
@Oberon00 Yes. We've had similar discussions about the interpretation when the same attribute key is set more than once on a span. The logical data model is that we have resources, distributed correlations, and span attributes. I would assert that span attributes override resources. I'm not sure whether span attributes or distributed correlations should take precedence.

tigrannajaryan commented:
> Yes. We've had similar discussions about the interpretation when the same attribute key is set more than once on a span. The logical data model is that we have resources, distributed correlations, and span attributes. I would assert that span attributes override resources. I'm not sure whether span attributes or distributed correlations should take precedence.

If we decide to do this I think it is important to have a language which says it is preferred to specify labels on the Resource if they are Span invariant. This is important from performance perspective.

SergeyKanzhelev commented:
I'm very sorry for bringing it late. But I think the same applies to `http.app_root`. I totally missed that there were two application attributes added. And I think they needs to be discussed together. Would you agree?

```
| `http.app_root` |The path prefix of the URL that identifies this HTTP application. If multiple roots exist, the one that was matched for this request should be used. | No |
```

Again, I'm very sorry for bringing it late

Oberon00 commented:
I did actually think about removing `http.app_root` along with `http.app`, but since even within a single app there can be multiple app roots, I maintain that even with a `http.app` in resources, it would make sense to have multiple app roots.
But I think that attribute falls firmly in the "nice-to-have" category, so if after reading that argument, you still think that the attribute should be removed, I can do that in the interest of getting this moved forward.

Oberon00 commented:
Also, if I follow @jmacd correctly, the decision whether some attribute should be put on resources would be demoted to an implementation detail/optimization decision that each instrumentation can make itself, which I would really like, it would make this discussion much easier.

SergeyKanzhelev commented:
@Oberon00 I feel strongly about it. I understand the need of the property, but, first, it may be a resource property and secondly it may not have an "http" prefix. Also not having it in this spec doesn't forbid to collect it.

*I tried to commit the change myself so I can merge it, but I don't have permissions*

Oberon00 commented:
> I tried to commit the change myself so I can merge it, but I don't have permissions

That's strange, I was sure that I ticked the "Allow maintainers to edit your PR" checkbox.

> first, it may be a resource property and secondly it may not have an "http" prefix

That seems to apply to `http.app` but not `http.app_root`. The application root is an URL path prefix which seems very HTTP specific to me. Also, it *might* be a resource if your app only has one. But I think specifying this as a resource would be overly restraining. E.g. consider having the same application `My nice WebShop` under both `shop.example.com/` (server_name = shop.example.com, app_root =/) and `example.com/shop` (server_name = example.com, app_root = /shop). Now assuming that resources are attached to the tracer (currently there is no API-level interface for resources, they are SDK-only so this would need more spec-work anyway), I'd need to do this (pseudo Python): 

```python

tracer= trace.get_factory().createTracer(
  'my.web.instrumentation',
  '1.2.3',
  extra_resources={
    'http.app_root': environ['CONTEXT_PREFIX'],
    'http.app': 'My cool web shop'
}

span = tracer.start_as_active_span(...)
span.set_attribute('http.server_name', environ['SERVER_NAME'])
```

And now I have a tracer+span that is created for only a single span. 😕

jmacd commented:
> And now I have a tracer+span that is created for only a single span. 😕

Do you see this as a problem? I have been thinking of the tracer returned in this case as merely a binding between the underlying SDK and some resource/component/service-name attributes.

SergeyKanzhelev commented:
This comment:

> The application root is an URL path prefix which seems very HTTP specific to me. Also, it might be a resource if your app only has one. 

Doesn't match to description in PR which says it "identifies the app":

>The path prefix of the URL that identifies this HTTP application. If multiple roots exist, the one that was matched for this request should be used.

In any case I'm happy to get into this discussion, but I really don't want to block this PR any further. I think it will be more productive to merge this without `app_root` and discuss app_root separately. Thank you!

Oberon00 commented:
@SergeyKanzhelev 

It still sounds to me like you talk about the `http.app`  attribute, not `http.app_root`. The `http.app` identifies the application, not the `http.app_root` (though the combination of `http.app_root` and `http.server_name` should also uniquely identify the app).

Anyway, I removed it for now, so I think this is now ready to be merged.

SergeyKanzhelev commented:
I was the last one blocking this PR. All concerns are now resolved. I'm merging now. Typically we would wait for some time after the last meaningful change. But in case of this long living PR I'd suggest to file an issue if this change was not aligned with what you have previously approved. Simply in interest of velocity
